src/core/meshcop/dataset_manager.cpp

```diff
@@ -788,6 +794,21 @@ void PendingDatasetManager::HandleDelayTimer(void)
             ExitNow();
         }
     }
+#if OPENTHREAD_CONFIG_AMEND_PARTIAL_PENDING_OPERATIONAL_DATASET
+    // If pending dataset does not contain all required data, merge missing Tlv's from active dataset
+    IgnoreError(Get<ActiveDatasetManager>().Read(activeDataset));
+    currentTlv = activeDataset.GetTlvsStart();
+    endTlv = activeDataset.GetTlvsEnd();
+    while(currentTlv < endTlv)
+    {
+        pendingTlv = dataset.GetTlv(currentTlv->GetType());
+        if(pendingTlv == nullptr || !Tlv::IsValid(*currentTlv))
+        {
+            dataset.SetTlv(*currentTlv);
+        }
+        currentTlv = currentTlv->GetNext();
+    }
+#endif
```

Comment:
How about this?  Tried to simplify the code
- Added check that given dataset does not have "NetworkKey" TLV
- Used `for` loop and defined the variable under the `if` block (so don;t need t define them at the top of methd)

```suggestion
#if OPENTHREAD_CONFIG_AMEND_PARTIAL_PENDING_OPERATIONAL_DATASET
    if (dataset.GetTlv(Tlv::kNetworkKey) == nullptr)
    {
        Dataset activeDataset;

        IgnoreError(Get<ActiveDatasetManager>().Read(activeDataset));

        for (Tlv *cur = activeDataset.GetTlvsStart(); cur < activeDataset.GetTlvsEnd(); cur = cur->GetNext())
        {
            if (dataset.GetTlv(cur->GetType()) == nullptr)
            {
                IgnoreError(dataset.SetTlv(*cur));
            }
        }
    }
#endif
```

Reply:
Thanks it looks great, but shouldn't we use Tlv::IsValid to validate the TLV just in case? or the TLv should never be invalid?

Reply:
I think we can assume TLVs would be valid. We are generally expected to accept pending dataset as is anyways.